### PR TITLE
Fix `unit.states.test_archive` for Windows

### PR DESCRIPTION
**Pull request by twangboy**
#### test_archive.py
Uses os agnostic paths
Mock `salt.utils.which` for Windows

#### archive.py
Handle urlparse inconsistencies on Windows

### What issues does this PR fix or reference?
https://github.com/saltstack/salt-jenkins/issues/439

### Tests written?
Yes

### CODE REVIEW

**Comment by twangboy** [comment]
Rebased...

**Comment by cachedout** [comment]
Bump @terminalmage

File: salt/states/archive.py [hunk 1]
```diff
@@ -765,12 +765,22 @@ def extracted(name,
...
     urlparsed_source = _urlparse(source_match)
-    source_hash_basename = urlparsed_source.path or urlparsed_source.netloc
+    urlparsed_scheme = urlparsed_source.scheme
+    urlparsed_path = os.path.join(
+        urlparsed_source.netloc,
+        urlparsed_source.path).rstrip(os.sep)
 
-    source_is_local = urlparsed_source.scheme in salt.utils.files.LOCAL_PROTOS
+    if urlparsed_scheme and \
+            urlparsed_scheme.lower() in 'abcdefghijklmnopqrstuvwxyz':
```
[the remark this answers is not in the document]
Comment (terminalmage) [reply salt/states/archive.py]: Additionally, this block of code could use a comment explaining that this check is to see if the source is a Windows path.

File: tests/unit/states/test_archive.py [hunk 2]
```diff
@@ -119,7 +127,7 @@ def test_tar_gnutar(self):
         Tests the call of extraction with gnutar
         '''
         gnutar = MagicMock(return_value='tar (GNU tar)')
-        source = '/tmp/foo.tar.gz'
+        source = os.sep + os.path.join('tmp', 'foo.tar.gz')
```
Comment (terminalmage) [inline tests/unit/states/test_archive.py]: Again, the `os.path.join()` usage seems unnecessary here. Does the test fail without it?

File: tests/unit/states/test_archive.py [hunk 3]
```diff
@@ -157,7 +166,7 @@ def test_tar_bsdtar(self):
         Tests the call of extraction with bsdtar
         '''
         bsdtar = MagicMock(return_value='tar (bsdtar)')
-        source = '/tmp/foo.tar.gz'
+        source = os.sep + os.path.join('tmp', 'foo.tar.gz')
```
Comment (terminalmage) [inline tests/unit/states/test_archive.py]: Same as the previous test. Does this one fail on Windows without `os.path.join()`?

**Review by terminalmage — CHANGES_REQUESTED** [review]

**Comment by twangboy** [comment]
Applied @terminalmage 's changes and rebased

**Comment by twangboy** [comment]
Rebased...
@terminalmage Could you approve?

**Review by terminalmage — APPROVED** [review]

**Pull request merged**